madmin/routes/apks.py

```diff
@@ -0,0 +1,105 @@
+
+    def allowed_file(self, filename):
+        return '.' in filename and filename.rsplit('.', 1)[1].lower() in global_variables.MAD_APK_ALLOWED_EXTENSIONS
+    
+    @auth_required
+    def mad_apks(self):
+        apks = utils.apk_util.get_mad_apks(self._db)
+        return render_template('madmin_apk_root.html', apks=apks)
+
+    @auth_required
+    def upload_file(self):
+        if request.method == 'POST':
+            try:
+                if len(request.files) == 0:
+                    flash('No selected file')
+                    return redirect(url_for('mad_apks'))
+                apk_upload = request.files['apk']
+                if apk_upload.filename == '':
+                    flash('No selected file')
+                    return redirect(url_for('mad_apks'))
+                if not self.allowed_file(apk_upload.filename):
+                    flash('File extension not allowed')
+                    return redirect(url_for('mad_apks'))
+                if apk_upload and self.allowed_file(apk_upload.filename):
+                    filename = secure_filename(apk_upload.filename)
+                    filepath = os.path.join(tempfile.gettempdir(), filename)
+                    apk_type = global_variables.MAD_APK_USAGE[request.form['apk_type']]
+                    apk_arch = global_variables.MAD_APK_ARCH[request.form['apk_arch']]
+                    # Temporarily save the file so we can do our version processing
+                    try:
+                        apk_upload.save(filepath)
+                        apk = apkutils.APK(filepath)
+                        version = apk.get_manifest()['@android:versionName']
+                        apk_upload.seek(0,0)
+                        data = apk_upload.read()
+                        logger.info('New APK uploaded for {} {} [{}]', request.form['apk_type'],
+                                                                       request.form['apk_arch'],
+                                                                       version)
+                        apk_upload.seek(0,0)
+                        # Determine if we already have this file-type uploaded.  If so, remove it once the new one is
+                        # completed and update the id
+                        filestore_id_sql = "SELECT `id` FROM `mad_apks` WHERE `usage` = %s AND `arch` = %s"
+                        filestore_id = self._db.autofetch_value(filestore_id_sql, args=(apk_type, apk_arch,))
+                        if filestore_id:
+                            delete_data = {
+                                'id': filestore_id
+                            }
+                            self._db.autoexec_delete('filestore', delete_data)
+                        insert_data = {
+                            'filename': filename,
+                            'data': apk_upload.read()
+                        }
+                        new_id = self._db.autoexec_insert('filestore', insert_data)

```

Comment:
What's the rationale for apk files being saved in the DB instead of being saved to disk directly?

On downside of writing large files to DB is that it will cause more issues with people asking for help on chaning their sql server settings.
additionally, have you done any testing for memory usage when serving files over DB?

Reply:
The reason was to support multi-instance deployments without having to upload the file to each instance.  In terms of memory it probably needs to be switched to a stream so it uses less.